### PR TITLE
Jdbc dbproc parser

### PR DESCRIPTION
**Pull request by val5244**
Pls, review

### CODE REVIEW

File: bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/DbprocJdbc.java [hunk 1]
```diff
@@ -1,7 +1,51 @@
...
 
+import java.io.BufferedReader;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+
+import org.pgcase.xobot.dbproc.jdbc.functions.JdbcFunctionExtractor;
+
 public class DbprocJdbc {
```
Comment (ruspl-afed) [inline bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/DbprocJdbc.java:10]: Please rework to remove code duplications

File: bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionExtractor.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import org.pgcase.xobot.dbproc.runtime.functions.XFunctionExtractor;
+
+public class JdbcFunctionExtractor implements XFunctionExtractor {
+	private static final String SCAN_DBPROC_FUNC_SQL_FILE = "/sql/dbproc_jdbc_functions.sql"; //$NON-NLS-1$
+	private static String sqlSentence = null;	
+
+	@Override
+	public Iterable<XFunctionDescriptor> extractFunctions(Object input, Map<String, Object> context, XIssueReporter reporter) {
+		if (input instanceof Connection) {
+			try (Connection jdbcConnection = (Connection) input) {
```
Comment (ruspl-afed) [inline bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionExtractor.java]: This will close the connection after the block, we agreed to not close it here

File: bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionExtractor.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+public class JdbcFunctionExtractor implements XFunctionExtractor {
+	private static final String SCAN_DBPROC_FUNC_SQL_FILE = "/sql/dbproc_jdbc_functions.sql"; //$NON-NLS-1$
+	private static String sqlSentence = null;	
+
+	@Override
+	public Iterable<XFunctionDescriptor> extractFunctions(Object input, Map<String, Object> context, XIssueReporter reporter) {
+		if (input instanceof Connection) {
+			try (Connection jdbcConnection = (Connection) input) {		
+				ArrayList<XFunctionDescriptor> functions = new ArrayList<XFunctionDescriptor>();
+				String schema = context.getOrDefault("schema", (Object)"public").toString();
```
Comment (ruspl-afed) [inline bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionExtractor.java]: String.valueOf(object) is safer than object.toString(). Also I would extract this to reusable utility like

public static String extractSchema(Map<String, Object> context)

File: bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionExtractor.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+					functions.add(parsed);
+				}
+				return functions;
+			} catch (SQLException e) {
+				String message = String.format("Failed to process input with context %s", context);
+				reporter.reportIssue(this, input, message, e);
+			} catch (IOException e1) {
+				String message = String.format("Failed to process input with context %s", context);
+				reporter.reportIssue(this, input, message, e1);
+			} catch (Exception e2) {
```
Comment (ruspl-afed) [inline bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionExtractor.java]: Probably the only block with general Exception is sufficient for now as you do not have specific processing in other blocks

File: bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionParser.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import org.pgcase.xobot.dbproc.runtime.functions.FunctionDescriptorImpl;
+import org.pgcase.xobot.dbproc.runtime.functions.XFunctionDescriptor;
+
+public class JdbcFunctionParser {
+
+	public static XFunctionDescriptor parse(Connection jdbcConnection, ResultSet resultSet, XIssueReporter reporter) 
+			throws SQLException 
+	{
+		
+		ArrayList<ArgumentDescriptorImpl> args = new ArrayList<ArgumentDescriptorImpl>();
```
Comment (ruspl-afed) [inline bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionParser.java]: The variable type should be declared via interface if possible, i.e. List<XArgumentDescriptor>

File: bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionParser.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+		ArrayList<String> options = new ArrayList<String>();
+		
+		XFunctionDescriptor desc = new FunctionDescriptorImpl(
```
Comment (ruspl-afed) [inline bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionParser.java]: please use FunctionBuilder, the FunctionDescriptorImpl will be inaccessible soon

File: bundles/org.pgcase.xobot.dbproc.runtime/src/org/pgcase/xobot/dbproc/runtime/triggers/TriggerDescriptorImpl.java [hunk 7]
```diff
@@ -1,13 +1,13 @@
...
-class TriggerDescriptorImpl implements XTriggerDescriptor {
+public class TriggerDescriptorImpl implements XTriggerDescriptor {
 
 	private final String object;
 	private final String name;
 	private final String action;
 	private final String events;
 
-	TriggerDescriptorImpl(String object, String name, String action, String events) {
+	public TriggerDescriptorImpl(String object, String name, String action, String events) {
```
Comment (ruspl-afed) [inline bundles/org.pgcase.xobot.dbproc.runtime/src/org/pgcase/xobot/dbproc/runtime/triggers/TriggerDescriptorImpl.java]: Please revert this change

File: bundles/org.pgcase.xobot.dbproc.runtime/src/org/pgcase/xobot/dbproc/runtime/triggers/TriggerDescriptorImpl.java [hunk 8]
```diff
@@ -1,13 +1,13 @@
 package org.pgcase.xobot.dbproc.runtime.triggers;
 
-class TriggerDescriptorImpl implements XTriggerDescriptor {
+public class TriggerDescriptorImpl implements XTriggerDescriptor {
```
Comment (ruspl-afed) [inline bundles/org.pgcase.xobot.dbproc.runtime/src/org/pgcase/xobot/dbproc/runtime/triggers/TriggerDescriptorImpl.java]: please revert this change and use TriggerBuilder

File: tests/org.pgcase.xobot.dbproc.jdbc.tests/OSGI-INF/l10n/bundle.properties [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+# Contributors:
+#     ArSysOp - initial API and implementation
+###############################################################################
+
+Bundle-Name = PgCase Xobot Parsers Tests
```
Comment (ruspl-afed) [inline tests/org.pgcase.xobot.dbproc.jdbc.tests/OSGI-INF/l10n/bundle.properties]: Please correct bundle name

File: tests/org.pgcase.xobot.dbproc.jdbc.tests/OSGI-INF/l10n/bundle.properties [hunk 10]
```diff
@@ -0,0 +1,25 @@
+#Properties file for org.pgcase.xobot.dbproc.antlr.tests
```
Comment (ruspl-afed) [inline tests/org.pgcase.xobot.dbproc.jdbc.tests/OSGI-INF/l10n/bundle.properties]: please correct bundle symbolic name

File: bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/triggers/JdbcTriggerParser.java [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+import org.pgcase.xobot.dbproc.runtime.XIssueReporter;
+import org.pgcase.xobot.dbproc.runtime.triggers.XTriggerDescriptor;
+import org.pgcase.xobot.dbproc.runtime.triggers.TriggerDescriptorImpl;
+
+public class JdbcTriggerParser {
+	public static XTriggerDescriptor parse(Connection jdbcConnection, ResultSet resultSet, XIssueReporter reporter) 
+			throws SQLException 
+	{
+		XTriggerDescriptor desc = new TriggerDescriptorImpl(
```
Comment (ruspl-afed) [inline bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/triggers/JdbcTriggerParser.java]: please replace with TriggerBuilder

File: bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/triggers/JdbcTriggerExtractor.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Map;
+
+import org.pgcase.xobot.dbproc.jdbc.DbprocJdbc;
+import org.pgcase.xobot.dbproc.runtime.XIssueReporter;
+import org.pgcase.xobot.dbproc.runtime.triggers.XTriggerDescriptor;
+import org.pgcase.xobot.dbproc.runtime.triggers.XTriggerExtractor;
+
+public class JdbcTriggerExtractor implements XTriggerExtractor {
```
Comment (ruspl-afed) [inline bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/triggers/JdbcTriggerExtractor.java:15]: Please apply here the review comments from JdbcFunctionExtractor

File: tests/org.pgcase.xobot.dbproc.jdbc.tests/src/org/pgcase/xobot/dbproc/jdbc/tests/SystemJdbcConnection.java [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+	public void close() {
+		try {
+			connection.createStatement().execute(sqlDropSchema);
+			connection.close();
+		} catch (SQLException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		} 
+		finally {
+			close();
```
Comment (ruspl-afed) [inline tests/org.pgcase.xobot.dbproc.jdbc.tests/src/org/pgcase/xobot/dbproc/jdbc/tests/SystemJdbcConnection.java]: Recursive in finally block seems dangerous.

File: tests/org.pgcase.xobot.dbproc.jdbc.tests/src/org/pgcase/xobot/dbproc/jdbc/tests/SimpleJdbcFunctionTest.java [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+import org.pgcase.xobot.dbproc.runtime.functions.XFunctionDescriptor;
+
+class SimpleJdbcFunctionTest {
+	
+	private static final String CREATE_TEST_FUNC = "sql/create_test_function1.sql"; //$NON-NLS-1$
+	private static SystemJdbcConnection conn; 
+	
+	@Test
+	void testFunction() throws IOException, SQLException {
+		conn = new SystemJdbcConnection();
```
Comment (ruspl-afed) [inline tests/org.pgcase.xobot.dbproc.jdbc.tests/src/org/pgcase/xobot/dbproc/jdbc/tests/SimpleJdbcFunctionTest.java]: Now the connection created here and closed somewhere in implementation
Let's try to implement more clear connection state management like
1) create
2) pass somebody to use
3) close

File: tests/org.pgcase.xobot.dbproc.jdbc.tests/src/org/pgcase/xobot/dbproc/jdbc/tests/SystemJdbcConnection.java [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+	        sqlCreateSchema = DbprocJdbc.getSqlFmt1(FILE_SQL_CREATE_SCHEMA,dbSchema);
+
+	        connection = DriverManager.getConnection(dbUrl,dbUser,dbPasswd);
+			Statement statement = connection.createStatement();
+			statement.execute(sqlDropSchema);
+			if (!connection.getAutoCommit()) connection.commit();
+			
+			statement = connection.createStatement();
+			statement.execute(sqlCreateSchema);
+			if (!connection.getAutoCommit()) connection.commit();
```
Comment (ruspl-afed) [inline tests/org.pgcase.xobot.dbproc.jdbc.tests/src/org/pgcase/xobot/dbproc/jdbc/tests/SystemJdbcConnection.java]: always use brackets

File: tests/org.pgcase.xobot.dbproc.jdbc.tests/src/org/pgcase/xobot/dbproc/jdbc/tests/SystemJdbcConnection.java [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+	        sqlDropSchema = DbprocJdbc.getSqlFmt1(FILE_SQL_DROP_SCHEMA,dbSchema);
+	        sqlCreateSchema = DbprocJdbc.getSqlFmt1(FILE_SQL_CREATE_SCHEMA,dbSchema);
+
+	        connection = DriverManager.getConnection(dbUrl,dbUser,dbPasswd);
+			Statement statement = connection.createStatement();
+			statement.execute(sqlDropSchema);
+			if (!connection.getAutoCommit()) connection.commit();
+			
+			statement = connection.createStatement();
+			statement.execute(sqlCreateSchema);
```
Comment (ruspl-afed) [inline tests/org.pgcase.xobot.dbproc.jdbc.tests/src/org/pgcase/xobot/dbproc/jdbc/tests/SystemJdbcConnection.java]: Possible statement resource leak after this line

File: tests/org.pgcase.xobot.dbproc.jdbc.tests/src/org/pgcase/xobot/dbproc/jdbc/tests/SystemJdbcConnection.java [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+	
+	public SystemJdbcConnection() throws SQLException, IOException {
+		try {
+	        sqlDropSchema = DbprocJdbc.getSqlFmt1(FILE_SQL_DROP_SCHEMA,dbSchema);
+	        sqlCreateSchema = DbprocJdbc.getSqlFmt1(FILE_SQL_CREATE_SCHEMA,dbSchema);
+
+	        connection = DriverManager.getConnection(dbUrl,dbUser,dbPasswd);
+			Statement statement = connection.createStatement();
+			statement.execute(sqlDropSchema);
+			if (!connection.getAutoCommit()) connection.commit();
```
Comment (ruspl-afed) [inline tests/org.pgcase.xobot.dbproc.jdbc.tests/src/org/pgcase/xobot/dbproc/jdbc/tests/SystemJdbcConnection.java]: always use brackets

**Review by ruspl-afed — CHANGES_REQUESTED** [review]

File: bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/DbprocJdbc.java (hunk 1 above)
Comment (val5244) [reply bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/DbprocJdbc.java:10]: Done

File: bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionExtractor.java (hunk 2 above)
Comment (val5244) [reply bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionExtractor.java]: jdbcConnection  - clone from input. Connection will not close, will lost context only. Original input will not touch except use for select.

Comment (ruspl-afed) [reply bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionExtractor.java]: You are using Java here and this is a reference. Try to debug to see what is going on there

Comment (val5244) [reply bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionExtractor.java]: I know about reference. I know that it's Java!
I debugged it. Connection created outside extractFunctions was exists and correct after extractFunctions exit. Really.
No need chjanges.

Your right. 
I had to redo it.

try {
    Connection jdbcConnection = (Connection) input;

cos i need close connection outside implementation, in test, and before close i need clear test schema.

File: bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionExtractor.java (hunk 3 above)
Comment (val5244) [reply bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionExtractor.java]: Extracted and moved. Done

File: bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionExtractor.java (hunk 4 above)
Comment (val5244) [reply bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionExtractor.java]: Done

File: bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionParser.java (hunk 5 above)
Comment (val5244) [reply bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionParser.java]: Done

File: bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionParser.java (hunk 6 above)
Comment (val5244) [reply bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionParser.java]: Done

File: bundles/org.pgcase.xobot.dbproc.runtime/src/org/pgcase/xobot/dbproc/runtime/triggers/TriggerDescriptorImpl.java (hunk 7 above)
Comment (val5244) [reply bundles/org.pgcase.xobot.dbproc.runtime/src/org/pgcase/xobot/dbproc/runtime/triggers/TriggerDescriptorImpl.java]: Done

File: bundles/org.pgcase.xobot.dbproc.runtime/src/org/pgcase/xobot/dbproc/runtime/triggers/TriggerDescriptorImpl.java (hunk 8 above)
Comment (val5244) [reply bundles/org.pgcase.xobot.dbproc.runtime/src/org/pgcase/xobot/dbproc/runtime/triggers/TriggerDescriptorImpl.java]: Done

File: tests/org.pgcase.xobot.dbproc.jdbc.tests/OSGI-INF/l10n/bundle.properties (hunk 9 above)
Comment (val5244) [reply tests/org.pgcase.xobot.dbproc.jdbc.tests/OSGI-INF/l10n/bundle.properties]: Bundle-Name = PgCase Xobot JDBC Parsers Tests
- this key only?

**Comment by val5244** [comment]
Please, review.
Test jdbc changed to heroku instance (credentials dont out to file yet).

File: bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionExtractor.java [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+			try  {		
+				Connection jdbcConnection = (Connection) input;
+				List<XFunctionDescriptor> functions = new ArrayList<XFunctionDescriptor>();
+				String schema = DbprocJdbc.extractSchema(context);						
+				if (sqlSentence ==null) {
+					sqlSentence = DbprocJdbc.getBundleSql(SCAN_DBPROC_FUNC_SQL_FILE);	
+				}
+				PreparedStatement preparedStatement = jdbcConnection.prepareStatement(sqlSentence);
+				preparedStatement.setString(1, schema);
+				ResultSet resultSet = preparedStatement.executeQuery();
```
Comment (ruspl-afed) [inline bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionExtractor.java:32]: Resource leak: PreparedStatement preparedStatement is not closed

File: bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionExtractor.java [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+					sqlSentence = DbprocJdbc.getBundleSql(SCAN_DBPROC_FUNC_SQL_FILE);	
+				}
+				PreparedStatement preparedStatement = jdbcConnection.prepareStatement(sqlSentence);
+				preparedStatement.setString(1, schema);
+				ResultSet resultSet = preparedStatement.executeQuery();			
+				while (resultSet.next()) {
+					XFunctionDescriptor parsed = JdbcFunctionParser.parse(jdbcConnection,resultSet,reporter);
+					functions.add(parsed);
+				}
+				return functions;
```
Comment (ruspl-afed) [inline bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/functions/JdbcFunctionExtractor.java:37]: Resource leak: ResultSet resultSet is not closed

File: bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/triggers/JdbcTriggerExtractor.java [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+		if (input instanceof Connection) {
+			try (Connection jdbcConnection = (Connection) input) {		
+				ArrayList<XTriggerDescriptor> triggers = new ArrayList<XTriggerDescriptor>();
+				String schema = context.getOrDefault("schema", (Object)"public").toString();
+				if (sqlSentence ==null) {
+					sqlSentence = DbprocJdbc.getBundleSql(SCAN_DBPROC_TRIG_SQL_FILE);	
+				}
+				PreparedStatement preparedStatement = jdbcConnection.prepareStatement(sqlSentence);
+				preparedStatement.setString(1, schema);
+				ResultSet resultSet = preparedStatement.executeQuery();
```
Comment (ruspl-afed) [inline bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/triggers/JdbcTriggerExtractor.java]: Resource leak: PreparedStatement preparedStatement is not closed

File: bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/triggers/JdbcTriggerExtractor.java [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+					sqlSentence = DbprocJdbc.getBundleSql(SCAN_DBPROC_TRIG_SQL_FILE);	
+				}
+				PreparedStatement preparedStatement = jdbcConnection.prepareStatement(sqlSentence);
+				preparedStatement.setString(1, schema);
+				ResultSet resultSet = preparedStatement.executeQuery();			
+				while (resultSet.next()) {
+					XTriggerDescriptor parsed = JdbcTriggerParser.parse(jdbcConnection,resultSet,reporter);
+					triggers.add(parsed);
+				}
+				return triggers;
```
Comment (ruspl-afed) [inline bundles/org.pgcase.xobot.dbproc.jdbc/src/org/pgcase/xobot/dbproc/jdbc/triggers/JdbcTriggerExtractor.java:37]: Resource leak: ResultSet resultSet is not closed

**Review by ruspl-afed — APPROVED** [review]
